Sources/MongoSwift/MongoCollection+BulkWrite.swift

```diff
@@ -0,0 +1,131 @@
+public struct BulkWriteOptions {
+    /// If `true`, when an insert fails, return without performing the remaining writes.
+    /// If `false`, when a write fails, continue with the remaining writes, if any.
+    /// Defaults to `true`.
+    public var ordered: Bool = true
+
+    /// If `true`, allows the write to opt-out of document level validation.
+    public let bypassDocumentValidation: Bool?
+}
+
+/// A protocol indicating write types that can be batched together using `MongoCollection.bulkWrite`.
+public protocol WriteModel {}
+
+/// A model for an `insertOne` command.
+public struct InsertOneModel: WriteModel {
+    /// The `Document` to insert.
+    public let document: Document
+}
+
+/// A model for a `deleteOne` command.
+public struct DeleteOneModel: WriteModel {
+    /// A `Document` representing the match criteria.
+    public let filter: Document
+
+    /// Specifies a collation to use.
+    public let collation: Document?
+}
+
+/// A model for a `deleteMany` command.
+public struct DeleteManyModel: WriteModel {
+    /// A `Document` representing the match criteria.
+    public let filter: Document
+
+    /// Specifies a collation to use.
+    public let collation: Document?
+}
+
+/// A model for a `replaceOne` command.
+public struct ReplaceOneModel: WriteModel {
+    /// A `Document` representing the match criteria.
+    public let filter: Document
+
+    /// The `Document` with which to replace the matched document.
+    public let replacement: Document
+
+    /// Specifies a collation to use.
+    public let collation: Document?
+
+    /// When `true`, creates a new document if no document matches the query.
+    public let upsert: Bool?
+}
+
+/// A model for an `updateOne` command.
+public struct UpdateOneModel: WriteModel {
+    /// A `Document` representing the match criteria.
+    public let filter: Document
+
+    /// A `Document` containing update operators.
+    public let update: Document
+
+    /// A set of filters specifying to which array elements an update should apply.
+    public let arrayFilters: [Document]?
+
+    /// Specifies a collation to use.
+    public let collation: Document?
+
+    /// When `true`, creates a new document if no document matches the query.
+    public let upsert: Bool?
+}
+
+/// A model for an `updateMany` command.
+public struct UpdateManyModel: WriteModel {
+    /// A `Document` representing the match criteria.
+    public let filter: Document
+
+    /// A `Document` containing update operators.
+    public let update: Document
+
+    /// A set of filters specifying to which array elements an update should apply.
+    public let arrayFilters: [Document]?
+
+    /// Specifies a collation to use.
+    public let collation: Document?
+
+    /// When `true`, creates a new document if no document matches the query.
+    public let upsert: Bool?
+}
+
+/// The result of a bulk write operation on a `MongoCollection`.
+public struct BulkWriteResult: Decodable {
```

Comment:
Does it make sense to provide accessors for write errors, write concern errors, and whether the write was acknowledged? PHP does that with [WriteResult](http://php.net/manual/en/class.mongodb-driver-writeresult.php).

If a bulk write execution results in an error, libmongoc will create an error message for us but users should still have a programmatic API to inspect the errors. This is particularly relevant if there are both write and write concern errors.

On a separate note, users may want to know whether a write was acknowledged as some fields should not be accessible otherwise (or they should be `nil`).

Reply:
given our conversation earlier, not totally sure how we want to handle this....
would it make sense to throw from  `MongoCollection.bulkWrite` if there is a write error or write concern error? that would mean if we actually get to the point of returning a `BulkWriteResult`, there are no errors.

Reply:
I think it's easier to stuff a single write result struct (derived from the libmongoc reply document) onto an error object than extra the count/ID fields into their own "result" object and split out the `writeErrors` array field and `writeConcernError` document field.

Reply:
> so I think for an unacknowledged write we should just return nil. we are already doing (or supposed to be doing) that in some other places in the driver, and since we have optionals I think it's more idiomatic.

If the Swift driver has no intention of reporting inserted IDs, I suppose `BulkWriteResult?` makes sensein place of a `BulkWriteResult` consisting entirely of unset optionals and a single `isAcknowledged = false` property.

I personally think it's useful for drivers to report inserted IDs, since those are often generated by the driver when a document without an `_id` field is inserted; however, a reasonable counter-argument is that any users that actually care about the inserted ID could simply set `_id` themselves before the insert operation.

> IMO it's kind of confusing to users if there's an error field when there doesn't need to be - they might think they need to manually check each BulkWriteResult they get back to see if writeErrors or writeConcernErrors have anything in them.

I think that `bulkWriteError` should still have a message and/or code. libmongoc is already going to provide a `bson_error_t` for us and we can use that as we normally do. The only extra behavior is that we also want to stuff the `bson_t` reply into the error as well.

One thought is that BulkWriteResult may need to be optional. Consider the scenario where an unacknowledged write encounters a network error. libmongoc is still going to fill out a `bson_error_t` even if there isn't anything to report in the `bson_t` reply.

> it does require a bit of work to split up the info, but I think our BSON decoder will automate some of that, and it saves users having to go down two nested levels (error -> result struct -> error fields) to actually see the errors.

I'm personally in favor of BulkWriteResult matching the `bson_t` reply provided by libmongoc, but if that complicates things I won't object. I imagine not splitting out the `writeErrors` and `writeConcernError` fields might prevent us from being able to re-use the same BulkWriteResult struct for both the successful `bulkWrite()` return value and the `bulkWriteError` type. In that case, let's aim for re-use.

Reply:
> I think that bulkWriteError should still have a message and/or code. libmongoc is already going to provide a bson_error_t for us and we can use that as we normally do. The only extra behavior is that we also want to stuff the bson_t reply into the error as well.
One thought is that BulkWriteResult may need to be optional. Consider the scenario where an unacknowledged write encounters a network error. libmongoc is still going to fill out a bson_error_t even if there isn't anything to report in the bson_t reply.

Ok, so how about this?

```swift
case writeError(code: Int32, message: String)
indirect case bulkWriteError(code: Int32, message: String, result: BulkWriteResult?, writeErrors: [MongoError], writeConcernError: Document?)
```

Adds a code and message to `bulkWriteError`, and makes the `BulkWriteResult` optional (network error case). I also changed `writeErrors` to actually be an array of error structs rather than just documents (added `indirect` keyword to support this nested structure).

Re: the inserted IDs and what `bulkWrite` returns - I don't have very strong feelings on what we do here. but I will note that we took the "no `acknowledged` field, just return `nil` if unacknowledged" approach for all the existing CRUD write ops, and in some of these cases, the optionally returned struct contains inserted IDs. I do agree that it's kind of odd to only provide the IDs when the write concern is acknowledged, since they exist regardless. 
I don't really care what we go with as long as we make the behavior consistent, possibly by modifying the existing write ops. @mbroadst did the initial CRUD API sketch so maybe he has opinions on this?

Reply:
> I also changed writeErrors to actually be an array of error structs rather than just documents (added indirect keyword to support this nested structure).

I think `writeConcernError` should also be an indirect case, too (assuming we have one for a write concern error).

Reply:
^ just updated my comment

Reply:
Should `writeErrors: [MongoError]` be `[MongoError]?`, or would you ensure that `writeErrors` is an empty array even if the only bulk write error was the `writeConcernError`?

Reply:
I was imagining we'd just use an empty array, but thinking about it more, I think making it an optional is better. it's nicer to be able to say like `if let writeErrs = error.writeErrors { ... }` instead of `if errors.writeErrors.count > 0 { ... }`.

 also saves us creating a bunch of unused arrays

Reply:
let's discuss the bulkWrite / unacknowledged stuff in triage as this is blocking further progress on our CRUD work

Reply:
We concluded to keep the existing insertOne/insertMany behavior and have operations return an optional result object (i.e. `nil` for `{w:0}`).